src/esssans/normalization.py

```diff
@@ -64,6 +66,92 @@ def solid_angle_rectangular_approximation(
     )
 
 
+def solid_angle(
+    data: CalibratedMaskedData[RunType],
+    pixel_shape: DetectorPixelShape[RunType],
+    transform: LabFrameTransform[RunType],
+) -> SolidAngle[RunType]:
+    """
+    Solid angle for cylindrical pixels.
+
+    Note that the approximation is valid when the distance from sample
+    to pixel is much larger than the length or the radius of the pixels.
+
+    Parameters
+    ----------
+    data:
+        The DataArray that contains the positions of the detector pixels in coords.
+        The positions must be in the sample reference frame.
+
+    pixel_shape:
+        Contains the description of the detector pixel shape.
+
+    transform:
+        Transformation from the local reference frame of the detector
+        to the sample reference frame.
+
+    Returns
+    -------
+    :
+        The solid angle of the detector pixels, as viewed from the sample position.
+    """
+    face_1_center, face_1_edge, face_2_center = (
+        pixel_shape['vertices']['vertex', i] for i in range(3)
+    )
+    cylinder_axis = transform * face_2_center - transform * face_1_center
+    radius = sc.norm(face_1_center - face_1_edge)
+    length = sc.norm(cylinder_axis)
+
+    return SolidAngle[RunType](
+        approximate_solid_angle_for_cylinder_shaped_pixel_of_detector(
+            pixel_position=data.coords['position'],
+            cylinder_axis=cylinder_axis,
+            radius=radius,
+            length=length,
+        )
+    )
+
+
+def approximate_solid_angle_for_cylinder_shaped_pixel_of_detector(
+    pixel_position: sc.Variable,
+    cylinder_axis: sc.Variable,
+    radius: sc.Variable,
+    length: sc.Variable,
+):
+    r"""Computes solid angle of a detector pixel under the assumption that the
+    distance to the detector pixel from the sample is large compared to the radius
+    and to the length of the piece of the detector cylinder that makes up the pixel.
+
+    The cylinder is approximated by a rectangle element contained in the cylinder.
+    The normal of the rectangle is selected to
+    1. be orthogonal to the cylinder axis and
+    2. "maximally parallel" to the pixel position vector.
+    This occurs when the normal :math:`n`, the cylinder axis :math:`c` and the
+    pixel position :math:`r` fall in a shared plane such that
+    .. math::
+        r = (n\cdot r)n + (c \cdot r)c.
+
+    From the above relationship we get the scalar product of the position
+    vector and the normal:
+
+    .. math::
+        |\hat{r}\cdot n| = \sqrt{\frac{r \cdot r - (c \cdot r^2)}{||r||^2}}
+
+    The solid angle contribution from the detector element is approximated as
+    .. math::
+        \Delta\Omega =  A \frac{|\hat{r}\cdot n|}{||r||^2}
+    where :math:`A = 2 R L` is the area of the rectangular element
+    and :math:`\hat{r}` is the normalized pixel position vector.
+    """
+    norm_pp = sc.norm(pixel_position)
+    norm_ca = sc.norm(cylinder_axis)
```

Comment:
The `cylinder_axis` is required to compute $cos \alpha$. Yes I guess it's length ought to be 1 but normalizing it makes sure it can't even become an issue if someone forgets to make sure it is length 1.

Reply:
Yes it is.